src/metabase/db/data_migrations.clj

```diff
@@ -172,6 +177,39 @@
                                  [:like
                                   :dashcard.visualization_settings "%\"click_link_template\":%"]]})))
 
+(defn- remove-group-id-from-mappings-by-setting-key
+   [mapping-setting-key admin-group-id]
+   ;; Intentionally get setting using `db/select-one-field` instead of `setting/get` because for some reasons
+   ;; during start-up setting/get return the default value defined in defsetting instead of value from Setting table
+   (let [mapping (try
+                  (json/parse-string (db/select-one-field :value Setting :key (name mapping-setting-key)))
+                  (catch Exception _e
+                    {}))]
+     (when-not (empty? mapping)
+       (setting/set-value-of-type!
+        :json mapping-setting-key
+        (into {}
+              (map (fn [[k v]] [k (filter #(not= admin-group-id %) v)]))
+              mapping)))))
+
+ (defmigration ^{:author "qnkhuat" :added "0.43.0"} migrate-remove-admin-from-group-mapping-if-needed
+   ;;  In the past we have a setting to disable group sync for admin group when using SSO or LDAP, but it's broken and haven't really worked (see #13820)
+   ;;  In #20991 we remove this option entirely and make sync for admin group just like a regular group.
+   ;;  But with this change we want to make sure we don't accidently add/remove adminusers we need to :
+   ;;  - for LDAP, if the `ldap-sync-admin-group` is disabled, remove all mapping for admin group
+   ;;  - for SAML, JWT remove all mapping for admin group
+   (let [admin-group-id (u/the-id (group/admin))]
+     (when (and (ldap/ldap-configured?)
+                (= "false" (db/select-one-field :value Setting :key "ldap-sync-admin-group")))
+       (remove-group-id-from-mappings-by-setting-key :ldap-group-mappings admin-group-id))
+     ;; sso are enterprise feature, so only remove admin groups if enterprise available
+     (when (premium-features/enable-sso?)
+       (classloader/require 'metabase-enterprise.sso.integrations.sso-settings)
+       (when ((resolve 'metabase-enterprise.sso.integrations.sso-settings/jwt-configured?))
+         (remove-group-id-from-mappings-by-setting-key :jwt-group-mappings admin-group-id))
```

Comment:
This is going to NPE on OSS.

```clj
(resolve 'this-does/not-exist)
=> nil

((resolve 'this-does/not-exist))
=>
;; Execution error (NullPointerException) at metabase.models.permissions/eval236040 (REPL:5).
;; Cannot invoke "clojure.lang.IFn.invoke()" because the return value of "clojure.lang.IFn.invoke(Object)" is null
```

Reply:
It'll not throw NPE because it is under `(when (premium-features/enable-sso?))` which only true if enterprise is available.

But Noah made [a good point](https://github.com/metabase/metabase/pull/20991#discussion_r833682196) about running it in OSS too,  so I'll change this in the way we could check if `smal/jwt` configured even in OSS

Reply:
I don't think copy-pasting all the implementations of those functions was really what I had in mind here. I just meant you should check that `(resolve 'metabase-enterprise.sso.integrations.sso-settings/jwt-configured?)` actual returns something before you invoke it as a function

Reply:
I think you're right, no need to check if enabled.
for saml and yml we haven't synced it, so it's safe to remove admin regardless.

but for ldap, we need to check if users have been syncing admin or not.